File: alexandria/ui/src/io/intino/alexandria/ui/displays/components/DateEditable.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.ui.displays.events.ChangeEvent;
import io.intino.alexandria.ui.displays.events.ChangeListener;

import java.time.Instant;

public class DateEditable<B extends Box> extends AbstractDateEditable<B> {
	private Instant min;
	private Instant max;
	private Instant value;
	private ChangeListener changeListener = null;

    public DateEditable(B box) {
        super(box);
    }

	public Instant value() {
		return value;
	}

	public DateEditable<B> value(Instant value) {
		this.value = value;
		return this;
	}

	public Instant min() {
		return this.min;
	}

	public DateEditable<B> min(Instant min) {
		this.min = min;
		return this;
	}

	public Instant max() {
		return this.max;
	}

	public DateEditable<B> max(Instant max) {
		this.max = max;
		return this;
	}

	public void update(Instant value) {
		value(value);
		notifier.refresh(value);
	}

	public DateEditable<B> onChange(ChangeListener listener) {
		this.changeListener = listener;
		return this;
	}

	public void notifyChange(Instant value) {
    	value(value);
    	if (changeListener != null) changeListener.accept(new ChangeEvent(this, value));
    }

}

File: alexandria/ui-elements/src/displays/components/DateEditable.js
import React from "react";
import { withStyles } from '@material-ui/core/styles';
import AbstractDateEditable from "../../../gen/displays/components/AbstractDateEditable";
import DateEditableNotifier from "../../../gen/displays/notifiers/DateEditableNotifier";
import DateEditableRequester from "../../../gen/displays/requesters/DateEditableRequester";
import MomentUtils from '@date-io/moment';
import { MuiPickersUtilsProvider, KeyboardDateTimePicker, KeyboardDatePicker } from '@material-ui/pickers';

const styles = props => ({
	date : {
		width: "100%"
	},
	datetime : {
		width: "100%"
	}
});

class DateEditable extends AbstractDateEditable {
	state = {
		value : this.props.value
	};

	constructor(props) {
		super(props);
		this.notifier = new DateEditableNotifier(this);
		this.requester = new DateEditableRequester(this);
	};

	handleChange(moment) {
		if (!moment.isValid()) return;
		this.requester.notifyChange(moment.toDate().getTime());
		this.setState({ value: moment.toDate() });
	};

	render() {
		const { min, max, timePicker, classes } = this.props;
		const dateLabel = this.translate("date");
		const timeLabel = this.translate("time");
		const pattern = this.props.pattern !== "" ? this.props.pattern : undefined;
		return (<div style={this.style()}>
					{ !timePicker ? <MuiPickersUtilsProvider utils={MomentUtils}><KeyboardDatePicker variant="inline" placeholder={pattern} autoOk
												format={pattern} className={classes.date} mask={this.props.mask}
												value={this.state.value} onChange={this.handleChange.bind(this)}
                                                minDate={min} maxDate={max} label={dateLabel}/></MuiPickersUtilsProvider> : undefined }
					{ timePicker ? <MuiPickersUtilsProvider utils={MomentUtils}><KeyboardDateTimePicker variant="inline" placeholder={pattern} autoOk
												   format={pattern} className={classes.datetime}
												   value={this.state.value} onChange={this.handleChange.bind(this)}
                                                   minDate={min} maxDate={max} label={timeLabel}/></MuiPickersUtilsProvider> : undefined }
				</div>
		);
	};

	refresh = (value) => {
		this.setState({ value: new Date(value) });
	};

}

export default withStyles(styles, { withTheme: true })(DateEditable);